### DISCUSSION TITLE
flphoto / Fl_Image_Display : fast image panning?

### DISCUSSION THREAD

**Discussion by fire-eggs** [fire-eggs/YAIV#24] [Ideas]
Back in 2005 timeframe, Michael Sweet wrote a program called flphoto based on FLTK. It includes a class called Fl_Image_Display which did a lot of the display work. The last, 1.3.1 source is at https://sourceforge.net/projects/fltk/files/flphoto/

I've not re-visited it lately, but when I first experimented with it, the code "felt" like it provided a faster panning experience than straight FLTK. This was on Windows, I've not yet tried it under Linux.

Fl_Image_Display uses a FLTK function called `fl_draw_image`. This function is different in that it takes a callback function. The callback function returns a single scanline from the image for display.

I think this optimizes display because only the _portion_ of the scanline required is returned. I.e. if the window is narrower than the image, only the visible viewport within the image is provided. This is in contrast with "typical" drawing, where AFAIK, the entire image is displayed, but clipped to the window viewport.

Under Windows, using GDI, I can easily believe this callback mechanism to be faster. I don't have a good sense of behavior under X11 or Wayland in Linux.

**Comment by rageworx** [comment]
Dear @fire-eggs ,
I was thought so in many years ago, but that callback type not recommend in modern FLTK as I know.
FLTK recommends implement Fl_Widget::draw() to user's widget class.

-- or --

If you want fast redraw mechanism on Windows, I recommend use like this,
```
#ifdef _WIN32
    InvalidateRect( fl_xid( window ), NULL, TRUE );
#endif /// of _WIN32
```
I'm using this way for rendering image over 60fps only for Windows with MSVC.

![image](https://user-images.githubusercontent.com/6948225/119433870-0c85df00-bd52-11eb-8134-487a7e3f6c31.png)

( This program is a part of fast rendering industrial camera viwer by using Pylon cameras over 60Hz/fps )

**Comment by rageworx** [comment]
### Testing with MSYS2 + MinGW-W64 ( 10.3.0, rev2 )

Primal test
- first , configure
- make
- holy ... a lot of errors on MinGW-W64 gcc/g++ ... are these very old C codes ?
- failure at all similar errors of " undefined reference to `__imp_ntohxxx`" ...

Fixing Makefile for MinGW-W64
- Appends fldcraw$(EXEEXT) to $(LIBS)
- error left for " undefined reference to `getc_unlocked' ", getc_unlocked declared in Linux pthread, not Windows.
- preprocessed,
```
#ifndef _WIN32
#define fgetc getc_unlocked
#endif
```
- no sys/wait.h in MinGW-W64
```
Compiling Fl_CRW_Image.cxx...
Fl_CRW_Image.cxx:36:10: fatal error: sys/wait.h: No such file or directory
   36 | #include <sys/wait.h>
```
- and these are for ...
```
Fl_CRW_Image.cxx: In constructor 'Fl_CRW_Image::Fl_CRW_Image(const char*)':
Fl_CRW_Image.cxx:61:16: error: 'fork' was not declared in this scope
   61 |     if ((pid = fork()) < 0)
      |                ^~~~
Fl_CRW_Image.cxx:75:11: error: 'waitpid' was not declared in this scope
   75 |       if (waitpid(pid, &status, 0) < 0)
```
- Sorry, fork() and waitpid() is not for Windows. ( it is not able to build, because Windows don't have fork() or pid. )

**Comment by rageworx** [comment]
## Testing on my embedded linux system ( Rockchip RK3399, my exclusive designed linux kernel 4.4.179, with X11 )
- compiler = gcc 6.3.0 20170516 ( debian 6.3.0-18-+deb9u1 ) aarch64
- ./configure , OK.
- make ... a lot of warnings ... wow ... is it Ok ?
- build failure at ...
```
Compiling Fl_CRW_Image.cxx...
In file included from /usr/include/string.h:630:0,
                 from flstring.h:33,
                 from pcd-file.c:22:
( ... many warnings ... )
pcd-file.c:82:33: warning: pointer targets in passing argument 2 of ‘__builtin_strcmp’ differ in signedness [-Wpointer-sign]
     if (0 == strncmp("PCD_OPA", img->mmap, 7)) {
                                 ^
( ... many warnings ... )

Compiling print.cxx...
Compiling slideshow.cxx...
Compiling espmsg.c...
Compiling testexif.cxx...
image.cxx: In member function ‘int flphoto::save_png(const char*, int)’:
image.cxx:650:33: error: ‘Z_BEST_COMPRESSION’ was not declared in this scope
   png_set_compression_level(pp, Z_BEST_COMPRESSION);
                                 ^~~~~~~~~~~~~~~~~~
Generating po/de...
make: ./espmsg: Command not found
Makefile:98: recipe for target 'po/de' failed
make: *** [po/de] Error 127
make: *** Waiting for unfinished jobs....
Makefile:95: recipe for target 'image.o' failed
make: *** [image.o] Error 1
fldcraw.c: In function ‘adobe_coeff’:
fldcraw.c:5431:16: warning: iteration 3 invokes undefined behavior [-Waggressive-loop-optimizations]
  cam_xyz[0][j] = table[i].trans[j] / 10000.0;
  ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
fldcraw.c:5430:7: note: within this loop
       for (j=0; j < 12; j++)
       ^~~
( ... many wrarnings ... )
fldcraw.c: In function ‘parse_phase_one’:
fldcraw.c:4884:19: warning: iteration 3 invokes undefined behavior [-Waggressive-loop-optimizations]
    romm_cam[0][i] = getreal(11);
    ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
fldcraw.c:4883:2: note: within this loop
  for (i=0; i < 9; i++)
  ^~~
```
- stopped by
```
Compiling testexif.cxx...
image.cxx: In member function ‘int flphoto::save_png(const char*, int)’:
image.cxx:650:33: error: ‘Z_BEST_COMPRESSION’ was not declared in this scope
   png_set_compression_level(pp, Z_BEST_COMPRESSION);
                                 ^~~~~~~~~~~~~~~~~~
```
- hmm ... #include <zlib.h> missed in image.cxx.
- then, successfully built all.
- run!

<img width="400" src="https://user-images.githubusercontent.com/6948225/119453290-3b13b200-bd72-11eb-9c13-9c8511c06579.png" />

<img width="400" src="https://user-images.githubusercontent.com/6948225/119454328-677bfe00-bd73-11eb-8be6-fdaa6376e644.png" />

<img width="400" src="https://user-images.githubusercontent.com/6948225/119454356-6d71df00-bd73-11eb-9c88-d23d4602cfef.png" />
( some images are not working )

- it's looks familiar ... where did I saw ?
- oh, I got it, I did made this similar type of viewer in 2009 ( https://rageworx.pe.kr/480?category=2427 , Korean )
- and GPI+ advanced version here ( https://rageworx.pe.kr/494?category=2427 , Korean )
- adn download availed oldest version is here, https://rageworx.pe.kr/477?category=2427
- RAWCONV you can test as like this, put directory image contains together with EXE, then drag directory to RAWCONV.
  ![image](https://user-images.githubusercontent.com/6948225/119454089-1704a080-bd73-11eb-91f5-de00e923dfd6.png)
- It still working but very buggy ... ( sorry, that was over 10 years ago )
  ![image](https://user-images.githubusercontent.com/6948225/119454218-461b1200-bd73-11eb-91b3-fbef0813149e.png)
- Anyway, this program is not handy ... is this you wanted ? I believe you didn't 😁

**Reply by fire-eggs** [reply]
>Anyway, this program is not handy ... is this you wanted ? I believe you didn't 

Ha ha! No, it's not! Just some history ...

FLTK's performance when panning isn't too bad, but I do see visual glitches. It may be my computers - they don't have GPUs ...

**Reply by rageworx** [reply]
> FLTK's performance when panning isn't too bad, but I do see visual glitches. It may be my computers - they don't have GPUs

Maybe it's some deficiencies of FLTK.
Basically, as I understood while using FLTK for years, it is not made for fast-updating interfaces like MS WPF or modern space-wasting-tinkering style app by using GPU.
But I thinking graphics viewer should enough to make it looks great with threading-lock-n-unlock to awake() with callback.
Or make looks better by using libfl_imgtk enhances 20 century style pixels ( I made fl_imgtk by this reason ).

Plus, There's many platform-dependency issues like MS' unclean GDI legacy support with modern high DPI with FLTK.
But I love to use FLTK for any other mostly-platform-indipendent GUI engines like QT, WxWidgets or something.
FLTK always helpful by less writing C/C++ codes for each different platforms except MS Windows ... ( always big issues was belong to MS Windows )

But I'd made many things to test graphics processing with FLTK and it's always so good to make it visible result with charm GUI.
I believe you can make a great YAIV soon.

Regards, Raph.
